File: swift/Tests/WarehouseTests/OrderTests.swift
import Testing
import Warehouse

@Suite("Order")
struct OrderTests {
  @Test func rejectsEmptyLineSet() {
    expectInvalidOrder(
      { _ = try Order(id: OrderId("ord-1"), lines: []) },
      contains: "at least one line"
    )
  }

  @Test func rejectsDuplicateSkus() {
    expectInvalidOrder(
      {
        _ = try Order(
          id: OrderId("ord-1"),
          lines: [line("SKU-1"), line("SKU-1")]
        )
      },
      contains: "duplicate"
    )
  }

  @Test func rejectsDuplicateSkusAfterNormalization() {
    expectInvalidOrder(
      {
        _ = try Order(
          id: OrderId("ord-1"),
          lines: [line("SKU-1"), line(" SKU-1 ")]
        )
      },
      contains: "duplicate"
    )
  }

  @Test func totalEqualsSumOfLineTotals() throws {
    let order = try Order(
      id: OrderId("ord-1"),
      lines: [
        line("SKU-1", quantity: 2, minorUnits: 500),
        line("SKU-2", quantity: 1, minorUnits: 1000),
      ]
    )
    #expect(try order.total() == try usd(2000))
  }

  @Test func rejectsMixedCurrenciesAtConstruction() {
    expectInvalidOrder(
      {
        _ = try Order(
          id: OrderId("ord-1"),
          lines: [
            line("SKU-1"),
            line("SKU-2", quantity: 1, minorUnits: 100, currency: "EUR"),
          ]
        )
      },
      contains: "mixed currencies"
    )
  }

  @Test func usesInjectedId() throws {
    let order = try Order(id: OrderId("ord-fixed-9"), lines: [line()])
    #expect(order.id.value == "ord-fixed-9")
    #expect(order.status == .new)
    #expect(order.version == 0)
  }

  @Test func rejectsEmptyOrderId() {
    expectInvalidOrder({ _ = try OrderId("   ") }, contains: "non-empty")
  }

  @Test func payTransitionsNewToPaid() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    try order.pay()
    #expect(order.status == .paid)
  }

  @Test func doublePayIsInvalid() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    try order.pay()
    expectInvalidOrder({ try order.pay() }, contains: "already been paid")
  }

  @Test func shipRequiresPaidState() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    expectInvalidOrder({ try order.ship() }, contains: "paid")
  }

  @Test func shipTransitionsPaidToShipped() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    try order.pay()
    try order.ship()
    #expect(order.status == .shipped)
  }

  @Test func payAfterShipRaisesOrderAlreadyShipped() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    try order.pay()
    try order.ship()
    do {
      try order.pay()
      Issue.record("expected OrderAlreadyShipped")
    } catch is OrderAlreadyShipped {
      // expected
    } catch {
      Issue.record("unexpected error \(error)")
    }
  }

  @Test func shipAfterShipRaisesOrderAlreadyShipped() throws {
    var order = try Order(id: OrderId("ord-1"), lines: [line()])
    try order.pay()
    try order.ship()
    do {
      try order.ship()
      Issue.record("expected OrderAlreadyShipped")
    } catch is OrderAlreadyShipped {
      // expected
    } catch {
      Issue.record("unexpected error \(error)")
    }
  }

  @Test func snapshotIsDetachedFromTheOriginal() throws {
    let order = try Order(id: OrderId("ord-1"), lines: [line()])
    var copy = order.snapshot()
    try copy.pay()
    copy.bumpVersion()
    #expect(order.status == .new)
    #expect(order.version == 0)
  }

  @Test func sequenceIdsAreNeverEmpty() {
    #expect(OrderId(sequence: 7).value == "ord-7")
  }
}

File: csharp/tests/Warehouse.UnitTests/OrderTests.cs
using Warehouse.Domain;

namespace Warehouse.UnitTests;

public sealed class OrderTests
{
    private static OrderId Id(string value = "ord-1") => new(value);

    private static OrderLine Line(string code = "SKU-1", int qty = 2, long minorUnits = 500) =>
        new(new Sku(code), new Quantity(qty), new Money(minorUnits, "USD"));

    [Fact]
    public void RejectsEmptyLineSet()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(
            () => new Order(Id(), Array.Empty<OrderLine>()));
        Assert.Contains("at least one line", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsDuplicateSkus()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(
            () => new Order(Id(), [Line("SKU-1"), Line("SKU-1")]));
        Assert.Contains("duplicate", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void RejectsDuplicateSkusAfterNormalization()
    {
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(
            () => new Order(Id(), [Line("SKU-1"), Line(" SKU-1 ")]));
        Assert.Contains("duplicate", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void TotalEqualsSumOfLineTotals()
    {
        var order = new Order(Id(), [Line("SKU-1", 2, 500), Line("SKU-2", 1, 1000)]);
        Assert.Equal(new Money(2000, "USD"), order.Total());
    }

    [Fact]
    public void RejectsMixedCurrenciesAtConstruction()
    {
        OrderLine[] mixed =
        [
            Line("SKU-1"),
            new OrderLine(new Sku("SKU-2"), new Quantity(1), new Money(100, "EUR")),
        ];
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(() => new Order(Id(), mixed));
        Assert.Contains("mixed currencies", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void UsesInjectedId()
    {
        var order = new Order(Id("ord-fixed-9"), [Line()]);
        Assert.Equal("ord-fixed-9", order.Id.Value);
        Assert.Equal(OrderStatus.New, order.Status);
        Assert.Equal(0, order.Version);
    }

    [Fact]
    public void PayTransitionsNewToPaid()
    {
        var order = new Order(Id(), [Line()]);
        order.Pay();
        Assert.Equal(OrderStatus.Paid, order.Status);
    }

    [Fact]
    public void DoublePayIsInvalid()
    {
        var order = new Order(Id(), [Line()]);
        order.Pay();
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(order.Pay);
        Assert.Contains("already been paid", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void ShipRequiresPaidState()
    {
        var order = new Order(Id(), [Line()]);
        InvalidOrderException error = Assert.Throws<InvalidOrderException>(order.Ship);
        Assert.Contains("paid", error.Message, StringComparison.Ordinal);
    }

    [Fact]
    public void ShipTransitionsPaidToShipped()
    {
        var order = new Order(Id(), [Line()]);
        order.Pay();
        order.Ship();
        Assert.Equal(OrderStatus.Shipped, order.Status);
    }

    [Fact]
    public void PayAfterShipRaisesOrderAlreadyShipped()
    {
        var order = new Order(Id(), [Line()]);
        order.Pay();
        order.Ship();
        Assert.Throws<OrderAlreadyShipped>(order.Pay);
    }

    [Fact]
    public void ShipAfterShipRaisesOrderAlreadyShipped()
    {
        var order = new Order(Id(), [Line()]);
        order.Pay();
        order.Ship();
        Assert.Throws<OrderAlreadyShipped>(order.Ship);
    }

    [Fact]
    public void SnapshotDoesNotAliasMutableState()
    {
        var order = new Order(Id(), [Line()]);
        Order snapshot = order.Snapshot();
        order.Pay();
        Assert.Equal(OrderStatus.New, snapshot.Status);
        Assert.Equal(OrderStatus.Paid, order.Status);
    }
}
